pipelinewise/cli/alert_sender.py

```diff
@@ -0,0 +1,129 @@
+"""
+PipelineWise CLI - Alert sender class
+"""
+import logging
+from typing import Dict
+from collections import namedtuple
+
+from .alert_handlers.base_alert_handler import BaseAlertHandler
+from .alert_handlers.slack_alert_handler import SlackAlertHandler
+
+from .alert_handlers.errors import InvalidAlertHandlerException
+from .alert_handlers.errors import NotImplementedAlertHandlerException
+from .alert_handlers.errors import NotConfiguredAlertHandlerException
+
+LOGGER = logging.getLogger(__name__)
+
+# Alert handler entries from PPW config.yml transformed to Tuples
+AlertHandler = namedtuple('AlertHandler', ['type', 'config'])
+
+# Register new alert handlers class here
+# The key is the alert handler name from the PPW config.yml
+# Every alert handler class needs to implement the BaseAlertHandler base class
+ALERT_HANDLER_TYPES_TO_CLASS = {
+    'slack': SlackAlertHandler
+}
+
+
+class AlertSender:
+    """
+    AlertDispatcher class
+
+    Takes a list of alert handlers that is normally defined
+    in the PPW config.yml and dispatching alert messages to
+    alert handler implementations
+    """
+
+    def __init__(self, alert_handlers: Dict = None) -> None:
+        # Initialise alert_handlers as empty dictionary if None provided
+        if not alert_handlers:
+            self.alert_handlers = dict()
+        else:
+            self.alert_handlers = alert_handlers
+
+        # Raise an exception if alert_handlers is not a dictionary
+        if not isinstance(self.alert_handlers, dict):
+            raise InvalidAlertHandlerException('alert_handlers needs to be a dictionary')
+
+    @staticmethod
+    def __init_handler_class(alert_handler: AlertHandler) -> BaseAlertHandler:
+        """
+        Takes an alert handler specification and initialises an alert handler class
+
+        Args:
+            alert_handler: an AlertHandler typed Tuple
+
+        Returns:
+            Initialised alert handler object
+        """
+        try:
+            # Get and initialise the correct alert handler class
+            alert_handler_class = ALERT_HANDLER_TYPES_TO_CLASS[alert_handler.type]
+            handler = alert_handler_class(alert_handler.config)
+        except KeyError:
+            raise NotImplementedAlertHandlerException(f'Alert handler type not implemented: {alert_handler.type}')
+
+        return handler
+
+    def __get_alert_handler(self, alert_handler_type: str) -> AlertHandler:
+        """
+        Get an alert handler from the alert handlers dict
+
+        Args:
+            alert_handler_type: type of the alert handler (slack, rollbar, etc.)
+
+        Returns:
+            AlertHandler tuple
+        """
+        if alert_handler_type in self.alert_handlers.keys():
```

Comment:
No need to call `.keys()`,  `alert_handlers` ia dictionary so `if alert_handler_type in self.alert_handlers:` is enough.

Reply:
fixed by https://github.com/transferwise/pipelinewise/pull/467/commits/d704908b6a2a80ee98b572d8ac47e8f94f0555af